### PR TITLE
MLDB-1508 json import improvements

### PR DESCRIPTION
**Pull request by jeremybarnes**
This removes many of the bottlenecks and inefficiencies in JSON importing.  It moves the speed up from parsing around 16,000 complex records per second to around 23,000.

### CODE REVIEW

**Comment by nicolaskruchten** [comment]
This is using `import.text` with `parse_json()` ?

**Comment by jeremybarnes** [comment]
No it's import.json

**Comment by nicolaskruchten** [comment]
oh... I was planning on retiring that procedure and recommending that we use import.text with parse_json() as * in the select parameter.

**Comment by nicolaskruchten** [comment]
Doesn't import.json basically just do that under the hood anyway?

**Comment by jeremybarnes** [comment]
Not really.  Import JSON has optimizations which will be harder to achieve with generic text import.  I agree with the goal though... In the end JSON is just text.

**Comment by jeremybarnes** [comment]
In any case, this PR doesn't modify the importers directly.  It improves the efficiency of the underlying mechanisms.

**Comment by nicolaskruchten** [comment]
Ok. I assumed that import.text would be faster because memory mapped..?
Does import.json do that trick too?

On Tuesday, 22 March 2016, Jeremy Barnes <notifications@github.com> wrote:

> In any case, this PR doesn't modify the importers directly. It improves
> the efficiency of the underlying mechanisms.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/mldbai/mldb/pull/278#issuecomment-200052199>
>


-- 
Nicolas Kruchten, Eng  |  Staff Software Engineer at Datacratic  |
http://datacratic.com/

**Comment by jeremybarnes** [comment]
Yes, that part of the machinery is common.  Although in the tests I'm running the file is compressed and so not memory mapped.

**Comment by nicolaskruchten** [comment]
Ok. And what is the performance loss from import.json to import.text?

On Tuesday, 22 March 2016, Jeremy Barnes <notifications@github.com> wrote:

> Yes, that part of the machinery is common. Although in the tests I'm
> running the file is compressed and so not memory mapped.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/mldbai/mldb/pull/278#issuecomment-200083294>
>


-- 
Nicolas Kruchten, Eng  |  Staff Software Engineer at Datacratic  |
http://datacratic.com/

**Comment by jeremybarnes** [comment]
We can test it.  It may not be much.  There are other performance improvements (sharing column names and other schema elements) that would be harder to do in the more generic context though not impossible.

Alternatively we could keep two implementations buy switch them based upon matching a pattern in the select expression.  That way at least it becomes an implementation detail.

File: sql/cell_value.cc [hunk 1]
```diff
@@ -1015,6 +1053,115 @@ CellValue::printInterval() const
...
+                context.newArrayElement();
+                context.writeInt(*p++);
+            }
+        }
+        context.endArray();
+        context.endObject();
+        return;
+    }
+    default:
+        throw HttpReturnException(400, "unknown cell type");
```
Comment (steadtler) [inline sql/cell_value.cc]: Code 500?

File: sql/coord.cc [hunk 2]
```diff
@@ -52,7 +52,15 @@ Coord(const char * str, size_t len)
 Coord::
 Coord(uint64_t i)
 {
-    initString(std::to_string(i));
+    // NOTE: std::to_string is SLOOOOW in GCC 4.8
```
Comment (steadtler) [inline sql/coord.cc]: maybe share with the code in cellvalue.cc using a template?

**Comment by steadtler** [comment]
Looks good, just a couple of suggestions but otherwise +1

File: sql/cell_value.cc (hunk 1 above)
Comment (jeremybarnes) [inline sql/cell_value.cc]: I'm leaving this one.  The code was simply moved.  There are lots of other places in this file that should probably be a 500; we can address them separately.

**Pull request merged**